src/analyses/base.ml

```diff
@@ -2286,27 +2322,34 @@ struct
        * variables of the called function from cpa_s. *)
       let add_globals (st: store) (fun_st: store) =
         (* Remove the return value as this is dealt with separately. *)
-        let cpa_noreturn = CPA.remove (return_varinfo ()) fun_st.cpa in
+        let cpa_noreturn =
+          if not longjmpthrough then
+            (* Remove the return value as this is dealt with separately. *)
+            CPA.remove (return_varinfo ()) fun_st.cpa
+          else
+            (* Keep the return value as this is not actually the return value but the thing supplied in longjmp *)
+            fun_st.cpa
```

Comment:
> Does this not mean that the RETURN variable remains in local state with the setjmp return value? Or why is it necessary to keep it like this?

Yes, that is precisely what we want here such that the correct value gets set here:

https://github.com/goblint/analyzer/blob/af4ce3b6bd85e1f62861906fbeb5ef4b53c0d53e/src/framework/constraints.ml#L694-L702

Reply:
It seems to me that snippet is calling `return` again (which adds the `RETURN` variable) and immediately `combine` without `~longjmpthrough` (which immediately removes the `RETURN` variable), or am I missing something?
Also, could it not be simpler if the `longjmp` value isn't passed off as `RETURN` but some separate `varinfo`, which `combine` leaves untouched? Then the above distinction wouldn't be required and `LONGJMP_RETURN` (or whatever) would simply be passed as any other variable.

On another note, is there a reason that `return` has to be called on the exceptional return state each time instead of storing the result of `return` at that new constraint unknown, just like normal function return nodes store the result of `return`?

Reply:
Yes, that is what's going on here: The situation here is the following: We are inside a function f that has performed a setjmp and called function g that did the longjmp.

Then, we do the following:
- Call `S.return` for this call to `g` (removes locals of g, and leaves the return value alone, as we pass `None` here)
- Call `S.combine` without `longjmpthrough` as we are in the context where the call is handled, this combines the local state and the callee state to obtain  the new local state, which then is side-effected to the appropriate location.

If the handling is not in the current context, the same thing happens except that we pass `longjmpthrough` in order to leave the return variable alone.

In principle, one could also do the `S.return` for the function itself in the handling of `longjmp` if it is not inside the current procedure and then always perform a return for the procedure if the target is not the caller in combine. I think I decided against it for symmetry reasons inside the handling of `longjmp` calls, but I agree that maybe that is actually the nicer way to structure it.